libraries/standard/http/src/http_client.c

```diff
@@ -29,23 +34,329 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+        if( ( pPath == NULL ) || ( pathLen == 0 ) )
+        {
+            /* Revise toAddLen to contain <HTTP_EMPTY_PATH_LEN> instead. */
+            toAddLen = ( toAddLen - pathLen ) + HTTP_EMPTY_PATH_LEN;
+            memcpy( pBufferCur, HTTP_EMPTY_PATH, HTTP_EMPTY_PATH_LEN );
+            pBufferCur += HTTP_EMPTY_PATH_LEN;
+        }
+        else
+        {
+            memcpy( pBufferCur, pPath, pathLen );
+            pBufferCur += pathLen;
+        }
+
+        memcpy( pBufferCur, SPACE_CHARACTER, SPACE_CHARACTER_LEN );
+        pBufferCur += SPACE_CHARACTER_LEN;
+
+        memcpy( pBufferCur,
+                HTTP_PROTOCOL_VERSION, HTTP_PROTOCOL_VERSION_LEN );
+        pBufferCur += HTTP_PROTOCOL_VERSION_LEN;
+        memcpy( pBufferCur,
+                HTTP_HEADER_LINE_SEPARATOR, HTTP_HEADER_LINE_SEPARATOR_LEN );
+        pRequestHeaders->headersLen += toAddLen;
+    }
+
+    return returnStatus;
+}
+
 /*-----------------------------------------------------------*/
 
 HTTPStatus_t HTTPClient_InitializeRequestHeaders( HTTPRequestHeaders_t * pRequestHeaders,
                                                   const HTTPRequestInfo_t * pRequestInfo )
 {
-    return HTTP_NOT_SUPPORTED;
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    /* Check for NULL parameters. */
+    if( pRequestHeaders == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders interface is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pRequestHeaders->pBuffer == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders->pBuffer is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( ( pRequestInfo == NULL ) )
+    {
+        IotLogError( "Parameter check failed: pRequestInfo interface is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( ( pRequestInfo->method == NULL ) )
+    {
+        IotLogError( "Parameter check failed: pRequestInfo->method is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pRequestInfo->pHost == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestInfo->pHost is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pRequestInfo->methodLen == 0 )
+    {
+        IotLogError( "Parameter check failed: pRequestInfo->methodLen must be greater than 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pRequestInfo->hostLen == 0 )
+    {
+        IotLogError( "Parameter check failed: pRequestInfo->hostLen must be greater than 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Empty else MISRA 15.7 */
+    }
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        /* Reset application-provided parameters. */
+        pRequestHeaders->headersLen = 0;
+        /* Clear user-provided buffer. */
+        memset( pRequestHeaders->pBuffer, 0, pRequestHeaders->bufferLen );
```

Comment:
A redundant write of the entire buffer is potentially a costly operation. Why is this necessary?